### ISSUE TITLE
Custom 50x error pages (due to large files)

### ISSUE THREAD

**Issue by Bjwebb** [OpenDataServices/cove#343]
Cove's behaviour changed on this last week, as we implemented per request limits on memory usage. https://github.com/OpenDataServices/cove/issues/210#issuecomment-215052958 This means the large request gets a "Bad Gateway" error instead of the server generally grinding to a halt.

We've had a support email about the Bad Gateway error page. In that case the user tried a smaller file which worked. It would be good to have a custom error page to suggest that if you used a large file, try a smaller one. Look like we should be able to set this up with an Apache `ErrorDocument 502` directive https://httpd.apache.org/docs/2.4/custom-error.html

**Comment by Bjwebb** [comment]
We should check what error messages our limits can current. When uploading a 100MB CSV I get a 503 Service Unavailable.

**Comment by caprenter** [comment]
@idlemoor will look into whether or not we can redirect this back to a django page
@caprenter could look at creating the page

**Comment by idlemoor** [comment]
https://httpd.apache.org/docs/2.4/custom-error.html

tl;dr @caprenter can create a nice page in django under /static. If you feel that the REDIRECT_URL and REDIRECT_QUERY_STRING environment variables (under "Available Variables") would be useful, please do use them.

And then I can add this to the Apache config
ErrorDocument 502 /static/dave's_nice_page

I think my side of this deal is entirely within the opendataservices-deploy repo, which is a nice bit of confirmation that we've got the boundaries in the right places.

You never know, it might just work. Step one would be to Just Do It on cove-dev with a really simple page. Let me know when you're up for it.

**Comment by Bjwebb** [comment]
I'm not sure that redirecting or forwarding this to a Django page is a good idea. These are also the errors we'll get if Django falls down completely.

Instead I think we want to create a simple standalone page, just a little more themed / friendly looking than the one we already have.

**Comment by caprenter** [comment]
If that's the technical solution - good. What are the steps to get this done then?

**Comment by Bjwebb** [comment]
Above, I said:
> just a little more themed / friendly looking than the one we already have.

Not sure I got that correct. Rereading this issue the key requirement is a page with some information on it about why we think a user might be seeing the error. Theming is a secondary concern.

My current opinion is that we shouldn't try to theme this too much, as 360/OCDS are different, and I'm not sure it's well spent effort.

Also, whilst redirecting to a dynamic Django page is a bad idea, creating a static file in our Django app is probably a good way to do this (as those are served by Apache). Therefore @idlemoor's suggestion above does make sense:

> And then I can add this to the Apache config
> ErrorDocument 502 /static/dave's_nice_page

Although probably with a different filename ;)

Next steps:
- [ ] Create a page with some appropriate words on in [cove/static](https://github.com/OpenDataServices/cove/tree/master/cove/static) - this will probably be me, since I have the best idea of what the words need to be.
- [ ] Modify the Apache config for dev

**Comment by Bjwebb** [comment]
I'm not sure why I thought this was 502 Bad Gateway, I can't see any instances of that in the apache logs.

What I do see is 500 errors, and a smaller number of 503 errors. A 500 error is what I get if I try to feed the big lottery data into cove dev.

**Comment by Bjwebb** [comment]
I've had a first go at setting up a custom 500 error page, with the message coming from a string in the config file:

![error](https://cloud.githubusercontent.com/assets/634/20107526/bc88d5a4-a5d0-11e6-8dfa-11efd8106e12.png)

**Comment by Bjwebb** [comment]
Maybe we should have some text along the lines of:

> Please file a GitHub issue or email code@opendataservices.coop if this problem persists.

**Comment by caprenter** [comment]
I think that is sensible - yes.

**Comment by Bjwebb** [comment]
I've updated the message to have a header, and with the above text, but it's still not styled in any way.

**Comment by Bjwebb** [comment]
We have a basic error message page now, and no more work on it is planned. If changes need to be made, e.g. to the appearance, those should be new issues.
